WIP: Caravan GUI 2.0

theodelrieu opened the pull request:
Hello,

This PR implements a full revamp of the current Caravan GUI. A deep rabbit hole it was.

### Rationale

When implementing fluid caravans (#304), I had to resort to a 0-sized inventory hack to be able to open the Caravan GUI.
Besides that, I feel that the current UI is quite clunky, which is not suprising given that you cannot anchor custom windows to "normal" ones.

I've tried my best to split the code in somewhat relevant subfolders, and even though I'm quite satisfied, there is still work to be done before removing the WIP tag.

I went with a different GUI update mechanism, based on my previous (and only) UI/UX experience in Flutter. I'm open to reverting to the previous one, let me know what you think of it.

I wanted to force the remote view, just like Trains UI, but the way caravan stations are selected doesn't work well with it (i.e. capsule-stick selection). It would be better if stations behaved more like train stations, but that's out of that PR's scope by a long shot, so I stopped here.

There are no tests, nor test plans attached, shame on me. I've tested quite extensively a while ago (came back from a two-month hiatus this week), but there is likely a lot of quirks lurking.

It's a huge single commit, apologies in advance...

Happy to discuss here or/and on Discord.

### Schedule + cargo tabs

<img width="448" height="918" alt="image" src="https://github.com/user-attachments/assets/17996139-cd76-4a36-8ae5-25a8d2f79c16" />
<img width="450" height="918" alt="image" src="https://github.com/user-attachments/assets/84564079-bea4-4cd7-b8a6-911efaf63219" />
<img width="448" height="102" alt="image" src="https://github.com/user-attachments/assets/eaf2bbb5-8ae7-4375-8f9d-07fa54390eb0" />

By far the most interesting part of the PR is shown on the second screen. I had to fully implement the inventory logic.
The full feature set, and the "won't dos" can be found at the top of `inventories.lua`.

Note that the "cargo" tab is greyed out when the caravan is out of the player's reach.

I've also redone the button + slider frame elements (c.f. third screen)

### Interrupts

<img width="454" height="914" alt="image" src="https://github.com/user-attachments/assets/f021287c-0c4d-45a5-968c-32440021e5c8" />
<img width="451" height="479" alt="image" src="https://github.com/user-attachments/assets/32c64076-a81d-42cf-a3fd-5169d64951eb" />
<img width="392" height="289" alt="image" src="https://github.com/user-attachments/assets/b9847077-b58e-4a55-a188-898d48f376c1" />

I went with two separate windows for add / edit interrupt, to mimick Vanilla's Train UI.
The Edit interrupt window operates on a temporary interrupt, to avoid modifying currently running ones and brick your base.
There is no "Discard changes" pop-up, so pressing escape will cancel every modification and you'll have to start over.
I went with that route because detecting / preventing clicks on the interface is very brittle, if at all possible.

notnotmelon commented:
Bug:

Use of a "char" destination type will self-delete on reaching the destination and print an error.

<img width="1301" height="664" alt="image" src="https://github.com/user-attachments/assets/73fc6485-4e59-489c-984b-814d9c2155cc" />

notnotmelon commented:
Bug; When moving an item with quality between caravan inventory and player inventory, the quality info is discarded and reverts to basic.

notnotmelon commented:
Bug; when ctrl+clicking to move all items into a caravan inventory, all power armors will lose their equipment grids.

theodelrieu commented:
Bug: mining a caravan crashes

theodelrieu commented:
>Bug; when ctrl+clicking to move all items into a caravan inventory, all power armors will lose their equipment grids.

Could not reproduce: filled a power armor with batteries, ctrl+clicked on an empty character inventory cell, batteries are still in the armor.

>Use of a "char" destination type will self-delete on reaching the destination and print an error.

Could not reproduce either, could you provide detailed steps? I've added my character as a destination but it didn't remove anything.

>Please increase the contrast on these icons

 Fixed in e69133a20614d67bde186c47ce4e80d1b2fe264b

>Currently the interrupt popup teleports to the top left corner after adding a new destination. Is there any way to prevent this?

Fixed in 3bd5e9efc3bfdd174bad1bf6281f4cb04d59a11e

theodelrieu commented:
>Bug; When moving an item with quality between caravan inventory and player inventory, the quality info is discarded and reverts to basic.

Fixed in 56d7c3ce5a9020bf9b3ddfa9058fa682744395f9

theodelrieu commented:
>Bug: When adding a new destination entity, an "Error" sound effect is played even though it worked successfully.

Seems to be already on master, reproduced by adding the spaceship as a destination.
Adding a random destination on the map didn't make any sound though.

theodelrieu commented:
>Use of a "char" destination type will self-delete on reaching the destination and print an error.

I rebased and can now see `Player: XXX` in the destination.

theodelrieu commented:
Fixed in 94c9de2b4fafb0c13caf5325d6b795ecedecc86a, but I didn't manage to show the equipment grid with `elem_tooltip`. Seems like a modding limitation. One also cannot open the armor with RMB, bummer.

<img width="350" height="334" alt="image" src="https://github.com/user-attachments/assets/f20b1156-fe5e-4f46-963a-b7d73fc1cb64" />

instead of

<img width="247" height="505" alt="image" src="https://github.com/user-attachments/assets/92861054-8b19-42f8-a84d-afc79e42c256" />

theodelrieu commented:
@notnotmelon I've fixed all the comments, went and rebased / split into several commits. They're not really atomic, but at least reviewable.

notnotmelon commented:
Bug: Clicking this button works, but does not update the GUI

![Uploading image.png…]()

notnotmelon commented:
Bug: Reassigning a "Destination is unavailable. Click here to reassign it." works, but does not automatically reopen the caravan GUI.

notnotmelon commented:
Bug: Shift+clicking an item from player inventory to caravan inventory does not update the player inventory GUI.
(Just a visual glitch, items are indeed moved properly)

theodelrieu commented:
>Bug: Reassigning a "Destination is unavailable. Click here to reassign it." works, but does not automatically reopen the caravan GUI.

Could reproduce on master. Both the remote view and carrot-capsule-stick stay up. Pressing Q makes the GUI reappears.

theodelrieu commented:
Regressions have been fixed and squashed
